### PR TITLE
feat(structured logs): converting logs to structured ones

### LINKED ISSUE #17: [PROPOSAL]  Structured log with -v flag
Converting actual logs to structured logs

```
* `--debug` This flag complete the logs with debug information (source file, line number).
* `--log-json` Set this flag to produce JSON formatted logs
```

### PR DESCRIPTION
**Pull request by CapKicklee**
Closes #17

### CODE REVIEW

File: internal/app/dataconnector/ping.go [hunk 1]
```diff
@@ -37,7 +38,7 @@ func newPingCommand(fullName string, err *os.File, out *os.File, in *os.File) *c
 		Run: func(cmd *cobra.Command, args []string) {
 			dc, e := dataconnector.Get(storage, args[0])
 			if e != nil {
-				logger.Error(fmt.Sprintf(e.Description))
+				log.Error().Msg(fmt.Sprintf(e.Description))
```
Comment (adrienaury) [inline internal/app/dataconnector/ping.go:41]:
```suggestion
				log.Error().Err(e).Msg(e.Description)
```

File: internal/app/pull/cli.go [hunk 2]
```diff
@@ -263,11 +258,11 @@ func (c epToStepListConverter) getRelation(name string) (pull.Relation, error) {
 	relation, ok := c.rmap[name]
 	if !ok {
 		err := fmt.Errorf("missing relation '%s' in relations.yaml", name)
-		logger.Error(err.Error())
+		log.Error().Msg(err.Error())
```
Comment (adrienaury) [inline internal/app/pull/cli.go:261]:
```suggestion
		log.Error().Err(err).Msg("")
```

File: internal/app/pull/http.go [hunk 3]
```diff
@@ -47,11 +48,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		for _, f := range strings.Split(query.Get("filter"), ",") {
 			kv := strings.SplitN(f, ":", 2)
 			if len(kv) != 2 {
-				logger.Error("can't parse filter\n")
+				log.Error().Msg("can't parse filter\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:51]:
```suggestion
				log.Error().Msg("can't parse filter")
```

File: internal/app/pull/http.go [hunk 4]
```diff
@@ -47,11 +48,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 				w.WriteHeader(http.StatusBadRequest)
 				_, ew := w.Write([]byte("{\"error\": \"param filter must be a string map (key1:value1,key2:value2)\"}\n"))
 				if ew != nil {
-					logger.Error("Write failed\n")
+					log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:55]:
```suggestion
					log.Error().Msg("Write failed")
```

File: internal/app/pull/http.go [hunk 5]
```diff
@@ -63,11 +64,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	if query.Get("limit") != "" {
 		limit64, elimit := strconv.ParseUint(query.Get("limit"), 10, 64)
 		if elimit != nil {
-			logger.Error("can't parse limie\n")
+			log.Error().Msg("can't parse limie\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:67]:
```suggestion
			log.Error().Msg("can't parse limit")
```

File: internal/app/pull/http.go [hunk 6]
```diff
@@ -63,11 +64,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 			w.WriteHeader(http.StatusBadRequest)
 			_, ew := w.Write([]byte("{\"error\" : \"param limit must be an positive integer\"}\n"))
 			if ew != nil {
-				logger.Error("Write failed\n")
+				log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:71]:
```suggestion
				log.Error().Msg("Write failed")
```

File: internal/app/pull/http.go [hunk 7]
```diff
@@ -82,35 +83,35 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 
 	if datasourceName, ok = pathParams["dataSource"]; !ok {
-		logger.Error("param datasource is required\n")
+		log.Error().Msg("param datasource is required\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:86]:
```suggestion
		log.Error().Msg("param datasource is required")
```

File: internal/app/pull/http.go [hunk 8]
```diff
@@ -82,35 +83,35 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 		w.WriteHeader(http.StatusBadRequest)
 		_, ew := w.Write([]byte("{\"error\": \"param datasource is required\"}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:90]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/pull/http.go [hunk 9]
```diff
@@ -82,35 +83,35 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 			return
 		}
 		return
 	}
 
 	datasource, err = getDataSource(datasourceName, w)
 	if err != nil {
-		logger.Error(err.Error())
+		log.Error().Msg(err.Error())
```
Comment (adrienaury) [inline internal/app/pull/http.go:98]:
```suggestion
		log.Error().Err(err).Msg("")
```

File: internal/app/pull/http.go [hunk 10]
```diff
@@ -82,35 +83,35 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 		w.WriteHeader(http.StatusNotFound)
 		_, ew := w.Write([]byte("{\"error\": \"" + err.Description + "\"}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:102]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/pull/http.go [hunk 11]
```diff
@@ -82,35 +83,35 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 			return
 		}
 		return
 	}
 
 	plan, e2 := getPullerPlan(filter, limit, where, idStorageFactory(query.Get("table")))
 	if e2 != nil {
-		logger.Error(e2.Error())
+		log.Error().Msg(e2.Error())
```
Comment (adrienaury) [inline internal/app/pull/http.go:110]:
```suggestion
		log.Error().Err(e2).Msg("")
```

File: internal/app/pull/http.go [hunk 12]
```diff
@@ -82,35 +83,35 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 		w.WriteHeader(http.StatusInternalServerError)
 		_, ew := w.Write([]byte("{\"error\": \"" + e2.Description + "}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:114]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/pull/http.go [hunk 13]
```diff
@@ -120,11 +121,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 
 	e3 := pull.Pull(plan, pull.NewOneEmptyRowReader(), datasource, pullExporter, pull.NoTraceListener{})
 	if e3 != nil {
-		logger.Error(e3.Error())
+		log.Error().Msg(e3.Error())
```
Comment (adrienaury) [inline internal/app/pull/http.go:124]:
```suggestion
		log.Error().Err(e3).Msg("")
```

File: internal/app/pull/http.go [hunk 14]
```diff
@@ -120,11 +121,11 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
 		w.WriteHeader(http.StatusInternalServerError)
 		_, ew := w.Write([]byte(e3.Description))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/pull/http.go:128]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/cli.go [hunk 15]
```diff
@@ -233,11 +228,11 @@ func (c idToPushConverter) getRelation(name string) push.Relation {
 
 	relation, ok := c.rmap[name]
 	if !ok {
-		logger.Error(fmt.Sprintf("missing relation %v in relations.yaml", name))
+		log.Error().Msg(fmt.Sprintf("missing relation %v in relations.yaml", name))
```
Comment (adrienaury) [inline internal/app/push/cli.go:231]:
```suggestion
		log.Error().Err(fmt.Errorf("missing relation %v in relations.yaml", name)).Msg("")
```

File: internal/app/push/http.go [hunk 16]
```diff
@@ -51,35 +52,35 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
 	)
 
 	if dcDestination, ok = pathParams["dataDestination"]; !ok {
-		logger.Error("param dataDestination is required\n")
+		log.Error().Msg("param dataDestination is required\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:55]:
```suggestion
		log.Error().Msg("param dataDestination is required")
```

File: internal/app/push/http.go [hunk 17]
```diff
@@ -51,35 +52,35 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 		w.WriteHeader(http.StatusBadRequest)
 		_, ew := w.Write([]byte("{\"error\": \"param dataDestination is required\"}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:59]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/http.go [hunk 18]
```diff
@@ -51,35 +52,35 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 			return
 		}
 		return
 	}
 
 	datadestination, err := getDataDestination(dcDestination)
 	if err != nil {
-		logger.Error(err.Error())
+		log.Error().Msg(err.Error())
```
Comment (adrienaury) [inline internal/app/push/http.go:67]:
```suggestion
		log.Error().Err(err).Msg("")
```

File: internal/app/push/http.go [hunk 19]
```diff
@@ -51,35 +52,35 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 		w.WriteHeader(http.StatusNotFound)
 		_, ew := w.Write([]byte("{\"error\": \"" + html.EscapeString(err.Description) + "\"}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:71]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/http.go [hunk 20]
```diff
@@ -51,35 +52,35 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 			return
 		}
 		return
 	}
 
 	plan, e2 := getPlan(idStorageFactory(query.Get("table")))
 	if e2 != nil {
-		logger.Error(e2.Error())
+		log.Error().Msg(e2.Error())
```
Comment (adrienaury) [inline internal/app/push/http.go:79]:
```suggestion
		log.Error().Err(e2).Msg("")
```

File: internal/app/push/http.go [hunk 21]
```diff
@@ -51,35 +52,35 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 		w.WriteHeader(http.StatusNotFound)
 		_, ew := w.Write([]byte("{\"error\": \"" + e2.Description + "\"}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:83]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/http.go [hunk 22]
```diff
@@ -88,11 +89,11 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
 	if query.Get("commitsize") != "" {
 		commitsize64, ecommitsize := strconv.ParseUint(query.Get("commitsize"), 10, 64)
 		if ecommitsize != nil {
-			logger.Error("can't parse commitsize\n")
+			log.Error().Msg("can't parse commitsize\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:92]:
```suggestion
			log.Error().Err(ecommitsize).Msg("can't parse commitsize")
```

File: internal/app/push/http.go [hunk 23]
```diff
@@ -88,11 +89,11 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 			w.WriteHeader(http.StatusBadRequest)
 			_, ew := w.Write([]byte("{\"error\" : \"param commitsize must be an positive integer\"}\n"))
 			if ew != nil {
-				logger.Error("Write failed\n")
+				log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:96]:
```suggestion
				log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/http.go [hunk 24]
```diff
@@ -104,33 +105,33 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
 		var edisableConstraints error
 		disableConstraints, edisableConstraints = strconv.ParseBool(query.Get("disable-constraints"))
 		if edisableConstraints != nil {
-			logger.Error("can't parse disable-constraints\n")
+			log.Error().Msg("can't parse disable-constraints\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:108]:
```suggestion
			log.Error().Err(edisableConstraints).Msg("can't parse disable-constraints")
```

File: internal/app/push/http.go [hunk 25]
```diff
@@ -104,33 +105,33 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 			w.WriteHeader(http.StatusBadRequest)
 			_, ew := w.Write([]byte("{\"error\" : \"param disable-constraints must be a boolean\"}\n"))
 			if ew != nil {
-				logger.Error("Write failed\n")
+				log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:112]:
```suggestion
				log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/http.go [hunk 26]
```diff
@@ -104,33 +105,33 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 		}
 	}
 
-	logger.Debug(fmt.Sprintf("call Push with mode %s", mode))
+	log.Debug().Msg(fmt.Sprintf("call Push with mode %s", mode))
 
 	e3 := push.Push(rowIteratorFactory(r.Body), datadestination, plan, mode, commitSize, disableConstraints, push.NoErrorCaptureRowWriter{})
 	if e3 != nil {
-		logger.Error(e3.Error())
+		log.Error().Msg(e3.Error())
```
Comment (adrienaury) [inline internal/app/push/http.go:123]:
```suggestion
		log.Error().Err(e3).Msg("")
```

File: internal/app/push/http.go [hunk 27]
```diff
@@ -104,33 +105,33 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 		w.WriteHeader(http.StatusNotFound)
 		_, ew := w.Write([]byte("{\"error\": \"" + e3.Description + "\"}"))
 		if ew != nil {
-			logger.Error("Write failed\n")
+			log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:127]:
```suggestion
			log.Error().Err(ew).Msg("Write failed")
```

File: internal/app/push/http.go [hunk 28]
```diff
@@ -104,33 +105,33 @@ func Handler(w http.ResponseWriter, r *http.Request, mode push.Mode) {
...
 			return
 		}
 		return
 	}
 	_, ew := w.Write([]byte("{\"error\": \"\"}"))
 	if ew != nil {
-		logger.Error("Write failed\n")
+		log.Error().Msg("Write failed\n")
```
Comment (adrienaury) [inline internal/app/push/http.go:134]:
```suggestion
		log.Error().Err(ew).Msg("Write failed")
```

File: internal/infra/pull/sql_datasource.go [hunk 29]
```diff
@@ -110,12 +111,12 @@ func (ds *SQLDataSource) RowReader(source pull.Table, filter pull.Filter) (pull.
 		fmt.Fprint(sql, ds.dialect.Limit(filter.Limit()))
 	}
 
-	if ds.logger != nil {
+	if log.Logger.GetLevel() <= zerolog.DebugLevel {
```
Comment (adrienaury) [inline internal/infra/pull/sql_datasource.go:114]: Have you check by debugging , if this test really return false if the zerolog logger is in info or upper level ?

File: pkg/dataconnector/driver.go [hunk 30]
```diff
@@ -17,13 +17,13 @@
...
 
-var logger Logger = Nologger{}
+import "github.com/rs/zerolog/log"
 
 // Add an alias to the storage, if it does not exist
 func Add(s Storage, m *DataConnector) *Error {
 	exist, err := Get(s, m.Name)
 	if err != nil {
-		logger.Error(err.Description)
+		log.Error().Msg(err.Description)
```
Comment (adrienaury) [inline pkg/dataconnector/driver.go:26]:
```suggestion
		log.Error().Err(err).Msg("")
```

Same with all other occurs

**Review by adrienaury — CHANGES_REQUESTED** [review]
Thank you for the pull request !

**Comment by adrienaury** [comment]
Similar to what is done with PIMO (https://github.com/CGI-FR/PIMO/tree/main/demo/demo9#fields) I think it would be nice to have a list of fields that are always printed when pulling/pushing.

Examples 

- id (or ids) of the current line being processed
- line number from the input (on push)
- line number from the output (on pull)
- ...

Another "nice to have" is a small result of the processing, for example with PIMO :

```
+-------+--------------------+--------------------+---------------------+--------+-------------+----------+------------+-------------+----------+
| level | stats:ignoredPaths | stats:skippedLines | stats:skippedFields | return | config      | duration | input-line | output-line | message  |
+-------+--------------------+--------------------+---------------------+--------+-------------+----------+------------+-------------+----------+
| info  | 0                  | 0                  | 0                   | 0      | masking.yml | 2.1671ms | 1          | 3           | End PIMO |
+-------+--------------------+--------------------+---------------------+--------+-------------+----------+------------+-------------+----------+
```

File: internal/infra/pull/sql_datasource.go (hunk 29 above)
Comment (CapKicklee) [reply internal/infra/pull/sql_datasource.go:114]: I'll add a test on this topic, checking if the log is here for an info level (if it's ok, then it should be for the upper levels)

Comment (CapKicklee) [reply internal/infra/pull/sql_datasource.go:114]: see https://github.com/CGI-FR/LINO/pull/18/commits/a53404cf61ec4c264d91ab4266f172bc2c4882b4

**Pull request merged**